tempodb/backend/azure/azure_helpers.go

```diff
@@ -59,88 +50,112 @@ func GetContainerURL(ctx context.Context, cfg *Config, hedge bool) (blob.Contain
 	if hedge && cfg.HedgeRequestsAt != 0 {
 		transport, stats, err = hedgedhttp.NewRoundTripperAndStats(cfg.HedgeRequestsAt, cfg.HedgeRequestsUpTo, transport)
 		if err != nil {
-			return blob.ContainerURL{}, err
+			return container.Client{}, err
 		}
 		instrumentation.PublishHedgedMetrics(stats)
 	}
 
-	client := http.Client{Transport: transport}
+	opts := azblob.ClientOptions{}
+	opts.Transport = &http.Client{Transport: transport}
+	opts.Retry = retry
+	opts.Telemetry = policy.TelemetryOptions{
+		ApplicationID: "Tempo",
+	}
 
-	httpSender := pipeline.FactoryFunc(func(next pipeline.Policy, po *pipeline.PolicyOptions) pipeline.PolicyFunc {
-		return func(ctx context.Context, request pipeline.Request) (pipeline.Response, error) {
-			// Send the request over the network
-			resp, err := client.Do(request.WithContext(ctx))
+	accountName := getStorageAccountName(cfg)
+	u, err := url.Parse(fmt.Sprintf("https://%s.%s", accountName, cfg.Endpoint))
 
-			return pipeline.NewHTTPResponse(resp), err
-		}
-	})
+	// If the endpoint doesn't start with blob. we can assume Azurite is being used
+	// So the endpoint should follow Azurite URL style
+	// https://learn.microsoft.com/en-us/rest/api/storageservices/get-blob#emulated-storage-service-uri
+	if !strings.HasPrefix(cfg.Endpoint, "blob.") {
+		u, err = url.Parse(fmt.Sprintf("http://%s/%s", cfg.Endpoint, accountName))
+	}
 
-	opts := blob.PipelineOptions{
-		Retry:      retryOptions,
-		Telemetry:  blob.TelemetryOptions{Value: "Tempo"},
-		HTTPSender: httpSender,
+	if err != nil {
+		return container.Client{}, err
 	}
 
-	if !cfg.UseFederatedToken && !cfg.UseManagedIdentity && cfg.UserAssignedID == "" {
-		credential, err := blob.NewSharedKeyCredential(getStorageAccountName(cfg), getStorageAccountKey(cfg))
+	var client *azblob.Client
+	accountKey := getStorageAccountKey(cfg)
+
+	switch {
```

Comment:
In my opinion, the most interesting part of the upgrade is the "new" authentication switch.

I'm not sure how to verify that the switch results in the same authentication behaviour as the old implementation. If I'm not mistaken, our end-to-end tests only cover shared key credential, and doesn't touch the other three authentication mechanisms at all.

Extra careful review might be in order here; any input is appreciated.

Reply:
I have basically no experience with Azure auth mechanisms. Unfortunately, we may just have to note this change loudly in the changelog and ask for quick feedback if operators see issues in azure.

If anyone knows how to write azure auth e2e tests it would be an excellent addition, but it's hard to throw resources down that rabbit hole.

Reply:
I took a brief look at whether we could set up some Azurite tests with different authentication mechanisms, but came up mostly empty-handed. It looks like Azurite [_might_](https://github.com/Azure/Azurite#default-storage-account) support some other authentication mechanisms, but I couldn't find documentation on how to set it up.

As a side note, the authentication switch currently assumes workload identity in the `default` case. After thinking about it some more, I think I prefer failing fast in case of invalid authentication configuration.

We have the `cfg.UseFederatedToken` option to explicitly specify workload identity authentication. A user who has accidentally misconfigured their authentication configuration might land in the `default` case and see an obscure error related to workload identity, when in reality their problem lies elsewhere.

I think I am in favour of returning an error in the `default` case. Something like this:
```go
	default:
		return container.Client{}, errors.New("please configure at least one Azure Blob Storage authentication mechanism")
```

Thoughts?

Reply:
I'm looking at our own azure configs and we don't set any of these flags. with the old way did it defafult to workload identity? if so we should probably keep it.

if we do keep default as workload identity can we use a fallthrough from the specific case to the default?

Reply:
> I'm looking at our own azure configs and we don't set any of these flags

I'm guessing you set `storage_account_name` and `storage_account_key`? I'd hazard that this is how most users authenticate.

There are five configuration values of interest when it comes to authentication:
* `use_managed_identity`
* `use_federated_token`
* `user_assigned_id`
* `storage_account_name`
* `storage_account_key`

If _none_ of these values are set, then:
* The **old** authentication logic would default to `NewSharedKeyCredential`, and fails immediately, _unless_ the `AZURE_STORAGE_ACCOUNT` and `AZURE_STORAGE_KEY` environment variables are set.
* The **new** authentication logic defaults to `NewWorkloadIdentityCredential`, and fails immediately _unless_ the `AZURE_CLIENT_ID`, `AZURE_TENANT_ID` and `AZURE_FEDERATED_TOKEN_FILE` environment variables are set.

There might be some users thet rely on the old default behaviour, so I'll update the new switch to honor that by defaulting to shared key credential.

Reply:
Action taken: the authentication switch now has three cases:
1. `case cfg.UseFederatedToken` covers workload identity.
2. `case cfg.UseManagedIdentity` covers both system-assigned and user-assigned managed identity.
3. `default` covers shared key credential.

This way, the default behaviour of the new authentication logic is the same as the old.

Reply:
Side note: Looks like use_federated_token is not mentioned in the current [documentation](https://grafana.com/docs/tempo/latest/configuration/azure/#azure-blob-storage-permissions-and-management).
Took me a while to figure out how to enable workload identity. This discussion actually made that possible for me.

Reply:
@LasseHels Thanks for keeping it consistent with the old behavior
@giskou Thank you for the heads up. I'm going to make a quick docs issue:

https://github.com/grafana/tempo/issues/2927